### PR TITLE
Adds Ability To Export Simulation Graph/Chart As PNG

### PR DESCRIPTION
**Pull request by RemingtonHolder**
### This PR adds the feature of exporting a simulation graph/chart as a PNG. 

The button for this is placed next to the zoom button at the bottom of the Simulation Graph/Chart Page.

_How I Did It:_ I simply reused some of the zoom button code to make a button, created a basic method employing JFileChooser so the user can pick where they want the image saved, and lastly used the JFreeChart -> ChartUtilities -> saveChartAsPNG method to export the chart as a PNG.

_Background:_ I was using this software for a project and was curious if this feature existed. I went to website and found that it was still on the Planned feature list but never seemed to be implemented, even in the 22.02 beta. I had a bit of spare time and decided that I would add the feature myself.

[**Link To Jar File For Testing (Hosted On OneDrive)**](https://1drv.ms/u/s!AmD6PPALFydSl3FVxPjJZaDnhNue)

**_Here Is An Example Exported Chart:_**
examplechartexport jpg

### CODE REVIEW

**Comment by hcraigmiller** [comment]
In so far as function is concerned, it works beautifully without anomalous behavior.

**Question**. The default directory for saving plots is the last directory from which a design file was **_navigated to_ and opened**, and, if you wish to save the plots in a different directory, _you must navigate to that directory each time a plot is saved_ (it does not remember where the last plot was saved. **Is this the desired behavior**, or should it remember the last directory in which a plot was saved?

In any event, well done.

**Comment by RemingtonHolder** [comment]
> In so far as function is concerned, it works beautifully without anomalous behavior.
> 
> 
> 
> **Question**. The default directory for saving plots is the last directory from which a design file was **_navigated to_ and opened**, and, if you wish to save the plots in a different directory, _you must navigate to that directory each time a plot is saved_ (it does not remember where the last plot was saved. **Is this the desired behavior**, or should it remember the last directory in which a plot was saved?
> 
> 
> 
> In any event, what you see is what you get. Well done.

I believe this is a weird oddity with JFileChooser, I'll look into it to see if this behavior may be changed and then do another PR

**Comment by JoePfeiffer** [comment]
You don't need to do another PR, you can push another commit to this one.

A very minor coding thing -- I believe all of our existing dialogs are full classes in .../swing/src/net/sf/openrocket/gui/dialogs/  Would it make sense for this to be as well?

**Comment by SiboVG** [comment]
I would make sure the File Format only has a 'PNG image (.png)' option, so remove the 'All files'-option and only use .png images in the 'Image files'-option. Currently it looks like this:
image

Whereas this is better, since you're only exporting PNG's:
image


I suggest you do this by creating a new translation entry on `line 99` of `core/resources/l10n/messages.properties`:
```XML
FileHelper.PNG_FILTER = PNG image (*.png)
```

Then in `swing/src/net/sf/openrocket/gui/util/FileHelper.java` on `line 58` you can add the new filter: 
```Java
/** File filter for PNG files (*.png) */
public static final FileFilter PNG_FILTER =
	new SimpleFileFilter(trans.get("FileHelper.PNG_FILTER"), ".png");
```

And then apply this new filter (+ remove the 'All files' option) in `SimulationPlotDialog.java` on `line 170`:
```Java
chooser.setAcceptAllFileFilterUsed(false);
chooser.setFileFilter(FileHelper.PNG_FILTER);
```

**Comment by RemingtonHolder** [comment]
I was going to do this but didn't know how to add to the translation file. I will add, thank you! :)> Cool functionality @RemingtonHolder, thanks!
> 
> 
> 
> But I would make sure the File Format only has a 'PNG image (.png)' option, so remove the 'All files'-option and only use .png images in the 'Image files'-option. Currently it looks like this:
> 
> image
> 
> 
> 
> Whereas this is better, since you're only exporting PNG's:
> 
> image
> 
> 
> 
> 
> 
> I suggest you do this by creating a new translation entry on `line 99` of `core/resources/l10n/messages.properties`:
> 
> ```XML
> 
> FileHelper.PNG_FILTER = PNG image (*.png)
> 
> ```
> 
> 
> 
> Then in `swing/src/net/sf/openrocket/gui/util/FileHelper.java` on `line 58` you can add the new filter: 
> 
> ```Java
> 
> /** File filter for PNG files (*.png) */
> 
> public static final FileFilter PNG_FILTER =
> 
> 	new SimpleFileFilter(trans.get("FileHelper.PNG_FILTER"), ".png");
> 
> ```
> 
> 
> 
> And then apply this new filter (+ remove the 'All files' option) in `SimulationPlotDialog.java` on `line 170`:
> 
> ```Java
> 
> chooser.setAcceptAllFileFilterUsed(false);
> 
> chooser.setFileFilter(FileHelper.PNG_FILTER);
> 
> ```

**Comment by RemingtonHolder** [comment]
Also does anyone know a method to get windows size so I can print the chart the size of it as displayed instead of just a fixed value?

**Comment by JoePfeiffer** [comment]
> Also does anyone know a method to get windows size so I can print the chart the size of it as displayed instead of just a fixed value?

I don't know the answer to your question, but if you're setting the size now I'd suggest letting the user do that would be nice.  If it is possible to get the original window size (and I can't imagine it's not, I've just never looked at that stuff) using that as the initial setting and letting the user change from there would be really nice.

**Comment by SiboVG** [comment]
> Also does anyone know a method to get windows size so I can print the chart the size of it as displayed instead of just a fixed value?

Change `line 168` to 
```Java
private boolean doPngExport(ChartPanel chartPanel, JFreeChart chart){
```
 Then on `line 188`:
```Java
ChartUtilities.saveChartAsPNG(file, chart, chartPanel.getWidth(), chartPanel.getHeight());
```
And on `line 124`:
```Java
doPngExport(chartPanel, jChart);
```

**Comment by SiboVG** [comment]
I'm also thinking about whether a 'Save'-icon would be better than the 'Print'-icon... @neilweinstock @hcraigmiller @JoePfeiffer any opinions on this?

**Comment by neilweinstock** [comment]
> I'm also thinking about whether a 'Save'-icon would be better than the 'Print'-icon... @neilweinstock @hcraigmiller @JoePfeiffer any opinions on this?

Seems like this is either a "save" or "export" function.  Our Export functionality is a bit of a jumble right now, but if this is confined to within the plot then it could either way.  I think it would be reasonable to label the button "Save as PNG" or "Save as Image".

**Comment by SiboVG** [comment]
@RemingtonHolder could you also take a look at the comments I left in your code (you can view them [here](https://github.com/openrocket/openrocket/pull/1253/files))

**Comment by RemingtonHolder** [comment]
@SiboVG I don't seem to be seeing them.?

**Comment by SiboVG** [comment]
> @SiboVG I don't seem to be seeing them.?

Ah no, sorry, you're right, I'll check back what I did wrong (I can see my comments, but you don't)

**Comment by SiboVG** [comment]
Should be online now (forgot to submit my review)

File: swing/src/net/sf/openrocket/gui/plot/SimulationPlotDialog.java [hunk 1]
```diff
@@ -117,11 +117,11 @@ public void actionPerformed(ActionEvent e) {
 		panel.add(button, "gapleft rel");
 
 		//// Print chart button
-		button = new SelectColorButton(Icons.FILE_PRINT);
+		button = new SelectColorButton("Save As Image");
```
[the remark this answers is not in the document]
Comment (RemingtonHolder) [reply swing/src/net/sf/openrocket/gui/plot/SimulationPlotDialog.java:120]: Ah yes, I forget that these dialogs have to be translated

File: swing/src/net/sf/openrocket/gui/plot/SimulationPlotDialog.java [hunk 2]
```diff
@@ -141,16 +158,42 @@ public void actionPerformed(ActionEvent e) {
...
-	
-	
-	
+
+	private boolean doPngExport(JFreeChart chart){
+		JFileChooser chooser = new JFileChooser();
+		chooser.setFileFilter(FileHelper.getImageFileFilter());
+		chooser.setCurrentDirectory(((SwingPreferences) Application.getPreferences()).getDefaultDirectory());
+
+		/* Ensures No Problems When Choosing File */
```
[the remark this answers is not in the document]
Comment (RemingtonHolder) [reply swing/src/net/sf/openrocket/gui/plot/SimulationPlotDialog.java:174]: I'm going to do `////` to make it uniform with the other single-line comments in this file

File: swing/src/net/sf/openrocket/gui/plot/SimulationPlotDialog.java [hunk 3]
```diff
@@ -141,16 +158,42 @@ public void actionPerformed(ActionEvent e) {
...
+		File file = chooser.getSelectedFile();
+		if (file == null)
+			return false;
+
+		file = FileHelper.forceExtension(file, "png");
+		if (!FileHelper.confirmWrite(file, this)) {
+			return false;
+		}
+
+		/* Uses JFreeChart Built In PNG Export Method */
```
[the remark this answers is not in the document]
Comment (RemingtonHolder) [reply swing/src/net/sf/openrocket/gui/plot/SimulationPlotDialog.java:187]: I'm going to do `////` to make it uniform with the other single-line comments in this file

File: core/resources/l10n/messages.properties [hunk 4]
```diff
@@ -1274,7 +1274,7 @@ TCMotorSelPan.btn.close = Close
 ! PlotDialog
 PlotDialog.CheckBox.Showdatapoints = Show data points
 PlotDialog.lbl.Chart = left click drag to zoom area. mouse wheel to zoom. ctrl-mouse wheel to zoom x axis only. ctrl-left click drag to pan.  right click drag to zoom dynamically.
-
+PlotDialog.btn.saveAsImage = Save As Image
```
Comment (SiboVG) [inline core/resources/l10n/messages.properties:1277]: I know I'm nitpicking, but best to add a whiteline (new line) after this, as to separate the PlotDialog translations and the ComponentTree translations

**Comment by JoePfeiffer** [comment]
> > I'm also thinking about whether a 'Save'-icon would be better than the 'Print'-icon... @neilweinstock @hcraigmiller @JoePfeiffer any opinions on this?
> 
> Seems like this is either a "save" or "export" function. Our Export functionality is a bit of a jumble right now, but if this is confined to within the plot then it could either way. I think it would be reasonable to label the button "Save as PNG" or "Save as Image".

After a brief perusal, I think we are consistently saying "Export" for everything other than .ork files.  So I'd like it to say "Export Image" and then make .png the only available file type.  We may want to export some other file format some day!

(I hope it doesn't feel like we're harassing you. We're very happy to see a new contributor, and want your work to slot as seamlessly as possible into the existing code)

**Comment by RemingtonHolder** [comment]
@JoePfeiffer No problem, I get that having application uniformity is important :)

**Comment by SiboVG** [comment]
Functionality and code looks good to me, thanks @RemingtonHolder

**Pull request merged**